stellargraph/mapper/graphwave_generator.py

```diff
@@ -59,86 +56,137 @@ def __init__(self, G, scales=(5, 10)):
                 full pass.
             num_parallel_calls (int): number of threads to use.
+            cache_filename (str): filename to store cached dataset.
         """
+
+        if not isinstance(batch_size, int):
+            raise TypeError(
+                f"batch_size: expected int, found {type(batch_size).__name__}"
+            )
+
+        if batch_size < 1:
+            raise ValueError(
+                f"batch_size: expected positive integer, found {batch_size}"
+            )
+
+        if not isinstance(num_parallel_calls, int):
+            raise TypeError(
+                f"num_parallel_calls: expected int, found {type(num_parallel_calls).__name__}"
+            )
+
+        if num_parallel_calls < 1:
+            raise ValueError(
+                f"num_parallel_calls: expected positive integer, found {num_parallel_calls}"
+            )
+
+        if not isinstance(shuffle, bool):
+            raise TypeError(f"shuffle: expected bool, found {type(shuffle).__name__}")
+
+        if not isinstance(repeat, bool):
+            raise TypeError(f"repeat: expected bool, found {type(repeat).__name__}")
+
         ts = tf.convert_to_tensor(sample_points.astype(np.float32))
 
-        dataset = tf.data.Dataset.from_tensor_slices(
-            self.eigen_vecs[self._node_lookup(node_ids)]
-        )
+        def _map_func(x):
+            return _empirical_characteristic_function(
+                _chebyshev(x, self.laplacian, self.coeffs, self.deg, self.max_eig), ts,
+            )
+
+        node_idxs = self._node_lookup(node_ids)
 
         # calculates the columns of U exp(-scale * eigenvalues) U^T on the fly
-        # and empirically the characteristic function for each column of U exp(-scale * eigenvalues) U^T
-        dataset = dataset.map(
-            lambda x: _empirical_characteristic_function(
-                tf.linalg.matvec(self.Ues, x), ts
-            ),
-            num_parallel_calls=num_parallel_calls,
-        )
+        # empirically calculate the characteristic function for each column of U exp(-scale * eigenvalues) U^T
 
-        if not targets is None:
+        dataset = tf.data.Dataset.from_tensor_slices(
+            tf.sparse.SparseTensor(
+                indices=np.stack([np.arange(len(node_ids)), node_idxs], axis=1),
+                dense_shape=(len(node_ids), self.laplacian.shape[0]),
+                values=np.ones(len(node_ids), dtype=np.float32),
+            )
+        ).map(_map_func, num_parallel_calls=num_parallel_calls)
 
-            target_dataset = tf.data.Dataset.from_tensor_slices(targets)
+        if targets is not None:
 
+            target_dataset = tf.data.Dataset.from_tensor_slices(targets)
             dataset = tf.data.Dataset.zip((dataset, target_dataset))
 
         # cache embeddings in memory for performance
+        dataset = dataset.cache(filename=cache_filename)
```

Comment:
I added it here for convenience; I was looking at running this on companies house 18 mil edges, 2mil nodes which was gonna take ~1hr to get embeddings for 1500 of the nodes. So I thought it would be convenient to run it once and save the embeddings and decided to add it in.

Agreed, its a little bit fiddly and might be best to leave it out. Plus users who understand how to use `.cache(filename='...')` properly can do this themselves anyway